File: AndroidPromotionSDK/src/main/java/com/ttcn/prm/ui/feature/mypromotion/MyPromotionViewModel.kt
package com.ttcn.prm.ui.feature.mypromotion

import com.ttcn.prm.ui.base.PRMStoreViewModel
import com.ttcn.prm.ui.feature.ext.toServiceSelectorUiItem
import com.ttcn.promotionsdk.config.PromotionSDKConfig
import com.ttcn.promotionsdk.domain.usecase.SearchCustomerVouchersUseCase
import com.ttcn.promotionsdk.presentation.mypromotion.MyPromotionIntent
import com.ttcn.promotionsdk.presentation.mypromotion.MyPromotionState
import com.ttcn.promotionsdk.presentation.mypromotion.MyPromotionStore
import com.ttcn.promotionsdk.presentation.serviceselector.servicesForApplicableProducts

/**
 * Màn "Ưu đãi của tôi". Nghiệp vụ nằm trọn ở [MyPromotionStore] (dùng chung với iOS); lớp này chỉ
 * giữ store sống qua xoay màn — xem [PRMStoreViewModel].
 *
 * Fragment đọc thẳng [MyPromotionState] và phát thẳng [MyPromotionIntent].
 */
internal class MyPromotionViewModel(
    searchCustomerVouchersUseCase: SearchCustomerVouchersUseCase,
    private val config: PromotionSDKConfig,
) : PRMStoreViewModel<MyPromotionState, MyPromotionIntent>(
    { scope -> MyPromotionStore(searchCustomerVouchersUseCase, scope) },
) {

    /** Dịch vụ cho bottom sheet "Chọn dịch vụ" — phần thuần Android, xem `PromotionDetailViewModel`. */
    fun serviceOptions(voucher: MyVoucherListItem): List<ServiceSelectorUiItem> =
        servicesForApplicableProducts(voucher.applicableProducts, config.availableServices)
            .map { it.toServiceSelectorUiItem() }
}

File: iosPromotionSDK/PromotionKit/MyPromotion/MyPromotionViewModel.swift
//
//  MyPromotionViewModel.swift
//  PromotionSDK
//
//  ĐỒNG NHẤT với `MyPromotionViewModel` bên Android — xem `PRMStoreViewModel` cho phần dùng chung.
//  Ở đây chỉ còn phần thuần iOS: điều hướng (router) và bottom sheet "Chọn dịch vụ".
//

import Foundation
@_implementationOnly import PRMKotlinBridge

final class MyPromotionViewModel: PRMScreenViewModel<MyPromotionRouter, MyPromotionStore> {

    /// Không nhận `DataModel`: màn này không có tham số đầu vào — token đọc từ
    /// `PromotionRequestContextProvider` của lõi.
    init(router: MyPromotionRouter,
         searchVouchersUseCase: SearchCustomerVouchersUseCase = SearchCustomerVouchersUseCase()) {
        super.init(router: router,
                   store: MyPromotionStore(searchCustomerVouchersUseCase: searchVouchersUseCase))
    }

    /// Mở Detail bằng voucher cơ bản; màn Detail tự fetch chi tiết (điều hướng — không ở store).
    func openDetail(voucherId: String) {
        guard let voucher = voucher(voucherId) else { return }
        router.routeToDetail(promotion: voucher.source)
    }

    /// iOS-only: Android mở màn Tìm kiếm bằng Fragment ở tầng view.
    func routeToSearch() {
        router.routeToSearch()
    }

    /// Đối ứng `MyPromotionViewModel.serviceOptions(voucher)` bên Android.
    func serviceOptions(voucherId: String) -> [ServiceSelectorItem] {
        guard let voucher = voucher(voucherId) else { return [] }
        return ServiceSelectorBuilder.items(forApplicableProducts: voucher.source.applicableProducts)
    }

    /// Tra voucher theo id — luật ở store (`MyPromotionState.voucher(id:)`), dùng chung với
    /// `SearchMyPromotionViewModel` (trước đây mỗi VM chép một bản y hệt).
    private func voucher(_ id: String) -> MyPromotionVoucher? {
        store.currentState().voucher(id: id)
    }
}
